shiny/ui/_markdown_stream.py

```diff
@@ -0,0 +1,276 @@
+            width=width,
+            height=height,
+        )
+
+    async def stream(
+        self,
+        content: Iterable[str] | AsyncIterable[str],
+        clear: bool = True,
+    ):
+        """
+        Stream content into the markdown stream.
+
+        Parameters
+        ----------
+        content
+            The content to stream. This can be any iterable of strings, such as a list,
+            generator, or file-like object.
+        clear
+            Whether to clear the existing content before streaming the new content.
+        """
+
+        content = _utils.wrap_async_iterable(content)
+
+        @reactive.extended_task
+        async def _task():
+            if clear:
+                await self._send_content_message("", "replace")
+
+            result = ""
+            async with self._streaming_dot():
+                async for c in content:
+                    result += c
+                    await self._send_content_message(c, "append")
+
+            return result
+
+        _task()
+
+        # Since the task runs in the background (outside/beyond the current context,
+        # if any), we need to manually raise any exceptions that occur
+        @reactive.effect
+        async def _handle_error():
+            e = _task.error()
+            if e:
+                await self._raise_exception(e)
+            _handle_error.destroy()  # type: ignore
+
+        return _task
+
+    async def clear(self):
+        """
+        Clear the content of the markdown stream.
+        """
+        return await self.stream([], clear=True)
+
+    @asynccontextmanager
+    async def _streaming_dot(self):
+        await self._send_stream_message(True)
+        try:
+            yield
+        finally:
+            await self._send_stream_message(False)
+
+    async def _send_content_message(
+        self,
+        content: str,
+        operation: Literal["append", "replace"],
+    ):
+        msg: ContentMessage = {
+            "id": self.id,
+            "content": content,
+            "operation": operation,
+        }
+        await self._send_custom_message(msg)
+
+    async def _send_stream_message(self, is_streaming: bool):
+        msg: isStreamingMessage = {
+            "id": self.id,
+            "isStreaming": is_streaming,
+        }
+        await self._send_custom_message(msg)
+
+    async def _send_custom_message(
+        self, msg: Union[ContentMessage, isStreamingMessage]
+    ):
+        if self._session.is_stub_session():
+            return
+        await self._session.send_custom_message("shinyMarkdownStreamMessage", {**msg})
+
+    async def _raise_exception(self, e: BaseException):
+        if self.on_error == "unhandled":
+            raise e
+        else:
+            sanitize = self.on_error == "sanitize"
+            msg = f"Error in MarkdownStream('{self.id}'): {str(e)}"
+            raise NotifyException(msg, sanitize=sanitize) from e
+
+
+@add_example()
+def markdown_stream_ui(
```

Comment:
Could we call this `output_markdown_stream()`? That feels like a better fit, or at least to fit with most user expectations. I think I'd rather `ui.output_markdown_stream()` than `ui.markdown_stream_ui()`.

Reply:
If we call it `output_markdown_stream()`, would you expect there to be a `@render.markdown_stream`?

Reply:
Sorry, I found https://github.com/posit-dev/py-shiny/pull/1782#discussion_r1936243033 after writing this.

But no, from a user perspective I don't think I'd find it weird. For me, the defining feature of an output is that I'm sending something from the server and it's showing up in the UI; the `@render` part seems incidental, like that's just how it works for plots or tables.

Reply:
Ok, that's fair, done in [a99b3a5](https://github.com/posit-dev/py-shiny/pull/1782/commits/a99b3a5bc4a0cdbc403a897f4de08de8473a90be).

It does make me wonder though -- should we change `chat_ui()` to `output_chat()` (for similar reasons)? I remember going back and forth and that one too and can't quite remember why I landed on `chat_ui()` (probably because it just feels weird for something that's technically an output in classical sense)

Reply:
> It does make me wonder though -- should we change `chat_ui()` to `output_chat()` (for similar reasons)?

I'd say the difference here is that Markdown Stream is unidirectional: it's only server → client. `chat_ui()` makes sense because there are also inputs: currently just the text input but it's very conceivable that we add additional input types to the UI. IOW, `chat_ui()` makes sense because it's not just an output.